include/revng/Pipeline/Step.h

```diff
@@ -0,0 +1,140 @@
+
+  /// Returns the predicted state of the Input containers status after the
+  /// execution of all the pipes in this step
+  ContainersStatus deduceResults(ContainersStatus Input) const;
+
+  /// Consumes the provided container and merges it into the current one
+  void mergeBack(ContainerSet RunningContainers) {
+    Containers.mergeBack(std::move(RunningContainers));
+  }
+
+  const ContainerSet &getContainers() const { return Containers; }
+
+  ContainerSet &getContainers() { return Containers; }
+
+  void addPipe(PipeWrapper Wrapper) { Pipes.push_back(std::move(Wrapper)); }
+
+  template<typename... PipeWrapperTypes>
+  Step(std::string Name,
+       ContainerSet Containers,
+       PipeWrapperTypes &&...PipeWrappers) :
+    Name(std::move(Name)),
+    Containers(std::move(Containers)),
+    Pipes({ std::forward<PipeWrapperTypes>(PipeWrappers)... }) {}
+
+  llvm::StringRef getName() const { return Name; }
+
+  template<typename OStream>
+  void dump(OStream &OS, size_t Indents = 0) const {
+    indent(OS, Indents);
+    OS << "Step " << Name << ":\n";
+
+    indent(OS, Indents + 1);
+    OS << "Pipes: \n";
+    for (const auto &Pipe : Pipes)
+      Pipe.dump(OS, Indents + 2);
+
+    indent(OS, Indents + 1);
+    OS << " containers: \n";
+    Containers.dump(OS, Indents + 2);
+  }
+
+  /// Drops from the backing containers all the targets presents in containers
+  /// status
+  llvm::Error invalidate(const ContainersStatus &ToRemove);
+
+  void dump() const debug_function { dump(dbg); }
+
+  llvm::Error storeToDisk(llvm::StringRef DirPath) const;
+  llvm::Error loadFromDisk(llvm::StringRef DirPath);
+
+  const ContainerBase *getContainer(llvm::StringRef ContainerName) const {
+    if (not Containers.contains(ContainerName))
+      return nullptr;
+    return &Containers.at(ContainerName);
+  }
+
+  ContainerBase *getOrCreateContainer(llvm::StringRef ContainerName) {
+    if (not Containers.containsOrCanCreate(ContainerName))
+      return nullptr;
+    return &Containers[ContainerName];
+  }
+
+private:
+  static void removeSatisfiedGoals(TargetsList &RequiredInputs,
+                                   const ContainerBase &CachedSymbols,
+                                   TargetsList &ToLoad);
+
+  void removeSatisfiedGoals(ContainersStatus &Targets,
+                            ContainersStatus &ToLoad) const;
+
+  void explainExecutedPipe(const PipeWrapper &Wrapper,
+                           llvm::raw_ostream *OS,
+                           size_t Indents = 0) const;
+  void explainStartStep(const ContainersStatus &Wrapper,
+                        llvm::raw_ostream *OS,
+                        size_t Indents = 0) const;
+};
+
+void prettyPrintTarget(const Target &Target,
+                       llvm::raw_ostream &OS,
+                       size_t Indents = 0);
+
+void prettyPrintStatus(const ContainersStatus &Targets,
+                       llvm::raw_ostream &OS,
+                       size_t Indents = 0);
```

Comment:
Are these declarations appropriate here?

Reply:
what do you mean? they are not debug stuff, they are not logically part of the class, and they are used in multiple cpp files. 
they are appropriate as free functions, i can moved them in another file but there is not much point to it.

Reply:
Wouldn't they be more appropriate in `Target.{h,cpp}` and wherever `ContainerStatus` is defined?